### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

### Questions
- Which factor influenced a candidate in getting placed?
- Does percentage matters for one to get placed?
- Which degree specialization is much demanded by corporate?
- Play with the data conducting all statistical tests.



- Berdasarkan status, faktor apa yang mempengaruhi kandidat untuk masuk.
- apakah persentase (ssc_p, hsc_p, degree_p, etest_p, dan mba_p) berpengaruh kepedasa seseorang untuk masuk>
- degree apa yang banyak diinginkan oleh perusahaan?

### Read The Data

In [2]:
data = pd.read_csv("Placement_Data_Full_Class.csv")
data.head()

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0      1      M  67.00   Others  91.00   Others  Commerce     58.00   
1      2      M  79.33  Central  78.33   Others   Science     77.48   
2      3      M  65.00  Central  68.00  Central      Arts     64.00   
3      4      M  56.00  Central  52.00  Central   Science     52.00   
4      5      M  85.80  Central  73.60  Central  Commerce     73.30   

    degree_t workex  etest_p specialisation  mba_p      status    salary  
0   Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1   Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2  Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3   Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4  Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0

In [3]:
data.describe()

sl_no       ssc_p       hsc_p    degree_p     etest_p       mba_p  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean   108.000000   67.303395   66.333163   66.370186   72.100558   62.278186   
std     62.209324   10.827205   10.897509    7.358743   13.275956    5.833385   
min      1.000000   40.890000   37.000000   50.000000   50.000000   51.210000   
25%     54.500000   60.600000   60.900000   61.000000   60.000000   57.945000   
50%    108.000000   67.000000   65.000000   66.000000   71.000000   62.000000   
75%    161.500000   75.700000   73.000000   72.000000   83.500000   66.255000   
max    215.000000   89.400000   97.700000   91.000000   98.000000   77.890000   

              salary  
count     148.000000  
mean   288655.405405  
std     93457.452420  
min    200000.000000  
25%    240000.000000  
50%    265000.000000  
75%    300000.000000  
max    940000.000000

In [4]:
data.isna().sum()

sl_no              0
gender             0
ssc_p              0
ssc_b              0
hsc_p              0
hsc_b              0
hsc_s              0
degree_p           0
degree_t           0
workex             0
etest_p            0
specialisation     0
mba_p              0
status             0
salary            67
dtype: int64

In [5]:
data_edited = data.copy()

In [6]:
data_edited.head()

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0      1      M  67.00   Others  91.00   Others  Commerce     58.00   
1      2      M  79.33  Central  78.33   Others   Science     77.48   
2      3      M  65.00  Central  68.00  Central      Arts     64.00   
3      4      M  56.00  Central  52.00  Central   Science     52.00   
4      5      M  85.80  Central  73.60  Central  Commerce     73.30   

    degree_t workex  etest_p specialisation  mba_p      status    salary  
0   Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1   Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2  Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3   Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4  Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0

In [7]:
data_edited.describe()

sl_no       ssc_p       hsc_p    degree_p     etest_p       mba_p  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean   108.000000   67.303395   66.333163   66.370186   72.100558   62.278186   
std     62.209324   10.827205   10.897509    7.358743   13.275956    5.833385   
min      1.000000   40.890000   37.000000   50.000000   50.000000   51.210000   
25%     54.500000   60.600000   60.900000   61.000000   60.000000   57.945000   
50%    108.000000   67.000000   65.000000   66.000000   71.000000   62.000000   
75%    161.500000   75.700000   73.000000   72.000000   83.500000   66.255000   
max    215.000000   89.400000   97.700000   91.000000   98.000000   77.890000   

              salary  
count     148.000000  
mean   288655.405405  
std     93457.452420  
min    200000.000000  
25%    240000.000000  
50%    265000.000000  
75%    300000.000000  
max    940000.000000

In [8]:
data_edited.head()

sl_no gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  \
0      1      M  67.00   Others  91.00   Others  Commerce     58.00   
1      2      M  79.33  Central  78.33   Others   Science     77.48   
2      3      M  65.00  Central  68.00  Central      Arts     64.00   
3      4      M  56.00  Central  52.00  Central   Science     52.00   
4      5      M  85.80  Central  73.60  Central  Commerce     73.30   

    degree_t workex  etest_p specialisation  mba_p      status    salary  
0   Sci&Tech     No     55.0         Mkt&HR  58.80      Placed  270000.0  
1   Sci&Tech    Yes     86.5        Mkt&Fin  66.28      Placed  200000.0  
2  Comm&Mgmt     No     75.0        Mkt&Fin  57.80      Placed  250000.0  
3   Sci&Tech     No     66.0         Mkt&HR  59.43  Not Placed       NaN  
4  Comm&Mgmt     No     96.8        Mkt&Fin  55.50      Placed  425000.0

Dropping sl_no because it has no correlation. As it stands for Serial Number

In [10]:
data_edited = data_edited.drop('sl_no',axis=1)

Mapping status, Placed = 1, Not Place = 0

In [9]:
data_edited['status'] = data_edited['status'].map({'Placed':1,'Not Placed':0})

Dropping salary column, because we want to find best factor that determine that someone get placed or not.

In [11]:
data_edited = data_edited.drop('salary',axis=1)

In [12]:
data_edited.head()

gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p   degree_t  \
0      M  67.00   Others  91.00   Others  Commerce     58.00   Sci&Tech   
1      M  79.33  Central  78.33   Others   Science     77.48   Sci&Tech   
2      M  65.00  Central  68.00  Central      Arts     64.00  Comm&Mgmt   
3      M  56.00  Central  52.00  Central   Science     52.00   Sci&Tech   
4      M  85.80  Central  73.60  Central  Commerce     73.30  Comm&Mgmt   

  workex  etest_p specialisation  mba_p  status  
0     No     55.0         Mkt&HR  58.80       1  
1    Yes     86.5        Mkt&Fin  66.28       1  
2     No     75.0        Mkt&Fin  57.80       1  
3     No     66.0         Mkt&HR  59.43       0  
4     No     96.8        Mkt&Fin  55.50       1

In [20]:
data_edited.describe()

ssc_p       hsc_p    degree_p     etest_p       mba_p      status
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000
mean    67.303395   66.333163   66.370186   72.100558   62.278186    0.688372
std     10.827205   10.897509    7.358743   13.275956    5.833385    0.464240
min     40.890000   37.000000   50.000000   50.000000   51.210000    0.000000
25%     60.600000   60.900000   61.000000   60.000000   57.945000    0.000000
50%     67.000000   65.000000   66.000000   71.000000   62.000000    1.000000
75%     75.700000   73.000000   72.000000   83.500000   66.255000    1.000000
max     89.400000   97.700000   91.000000   98.000000   77.890000    1.000000

In [13]:
data_edited['specialisation'][data_edited['status'] == 1].value_counts()

Mkt&Fin    95
Mkt&HR     53
Name: specialisation, dtype: int64

- Mkt&Fin banyak dicari oleh perusahaan, dengan perbandingan 95/148. 53 lainnya mengambil spesialisasi mkt&hr.

In [21]:
data_edited['specialisation'] = data_edited['specialisation'].map({'Mkt&Fin' : 0, 'Mkt&HR' : 1})

In [22]:
data_edited.head()

gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p   degree_t  \
0      M  67.00   Others  91.00   Others  Commerce     58.00   Sci&Tech   
1      M  79.33  Central  78.33   Others   Science     77.48   Sci&Tech   
2      M  65.00  Central  68.00  Central      Arts     64.00  Comm&Mgmt   
3      M  56.00  Central  52.00  Central   Science     52.00   Sci&Tech   
4      M  85.80  Central  73.60  Central  Commerce     73.30  Comm&Mgmt   

  workex  etest_p  specialisation  mba_p  status  
0     No     55.0               1  58.80       1  
1    Yes     86.5               0  66.28       1  
2     No     75.0               0  57.80       1  
3     No     66.0               1  59.43       0  
4     No     96.8               0  55.50       1

In [14]:
data_edited['degree_t'][data_edited['status'] == 1].value_counts()

Comm&Mgmt    102
Sci&Tech      41
Others         5
Name: degree_t, dtype: int64

In [23]:
data_edited['degree_t'] = data_edited['degree_t'].map({'Comm&Mgmt' : 0, 'Sci&Tech' : 1, 'Others' : 2})

In [24]:
data_edited.head()

gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  degree_t workex  \
0      M  67.00   Others  91.00   Others  Commerce     58.00         1     No   
1      M  79.33  Central  78.33   Others   Science     77.48         1    Yes   
2      M  65.00  Central  68.00  Central      Arts     64.00         0     No   
3      M  56.00  Central  52.00  Central   Science     52.00         1     No   
4      M  85.80  Central  73.60  Central  Commerce     73.30         0     No   

   etest_p  specialisation  mba_p  status  
0     55.0               1  58.80       1  
1     86.5               0  66.28       1  
2     75.0               0  57.80       1  
3     66.0               1  59.43       0  
4     96.8               0  55.50       1

In [15]:
data_edited['gender'].value_counts()

M    139
F     76
Name: gender, dtype: int64

In [25]:
data_edited['gender'] = data_edited['gender'].map({'M' : 0, 'F' : 1})

In [26]:
data_edited.head()

gender  ssc_p    ssc_b  hsc_p    hsc_b     hsc_s  degree_p  degree_t  \
0       0  67.00   Others  91.00   Others  Commerce     58.00         1   
1       0  79.33  Central  78.33   Others   Science     77.48         1   
2       0  65.00  Central  68.00  Central      Arts     64.00         0   
3       0  56.00  Central  52.00  Central   Science     52.00         1   
4       0  85.80  Central  73.60  Central  Commerce     73.30         0   

  workex  etest_p  specialisation  mba_p  status  
0     No     55.0               1  58.80       1  
1    Yes     86.5               0  66.28       1  
2     No     75.0               0  57.80       1  
3     No     66.0               1  59.43       0  
4     No     96.8               0  55.50       1

In [16]:
data_edited['ssc_b'].value_counts()

Central    116
Others      99
Name: ssc_b, dtype: int64

In [27]:
data_edited['ssc_b'] = data_edited['ssc_b'].map({'Central' : 0, 'Others' : 1})

In [28]:
data_edited.head()

gender  ssc_p  ssc_b  hsc_p    hsc_b     hsc_s  degree_p  degree_t workex  \
0       0  67.00      1  91.00   Others  Commerce     58.00         1     No   
1       0  79.33      0  78.33   Others   Science     77.48         1    Yes   
2       0  65.00      0  68.00  Central      Arts     64.00         0     No   
3       0  56.00      0  52.00  Central   Science     52.00         1     No   
4       0  85.80      0  73.60  Central  Commerce     73.30         0     No   

   etest_p  specialisation  mba_p  status  
0     55.0               1  58.80       1  
1     86.5               0  66.28       1  
2     75.0               0  57.80       1  
3     66.0               1  59.43       0  
4     96.8               0  55.50       1

In [29]:
data_edited['hsc_b'].value_counts()

Others     131
Central     84
Name: hsc_b, dtype: int64

In [30]:
data_edited['hsc_b'] = data_edited['hsc_b'].map({'Others' : 0, 'Central' :1})

In [31]:
data_edited.head()

gender  ssc_p  ssc_b  hsc_p  hsc_b     hsc_s  degree_p  degree_t workex  \
0       0  67.00      1  91.00      0  Commerce     58.00         1     No   
1       0  79.33      0  78.33      0   Science     77.48         1    Yes   
2       0  65.00      0  68.00      1      Arts     64.00         0     No   
3       0  56.00      0  52.00      1   Science     52.00         1     No   
4       0  85.80      0  73.60      1  Commerce     73.30         0     No   

   etest_p  specialisation  mba_p  status  
0     55.0               1  58.80       1  
1     86.5               0  66.28       1  
2     75.0               0  57.80       1  
3     66.0               1  59.43       0  
4     96.8               0  55.50       1

In [17]:
data_edited['hsc_s'].value_counts()

Commerce    113
Science      91
Arts         11
Name: hsc_s, dtype: int64

In [32]:
data_edited['hsc_s'] = data_edited['hsc_s'].map({'Commerce' : 0, 'Science' : 1, 'Arts' : 2})

In [33]:
data_edited.head()

gender  ssc_p  ssc_b  hsc_p  hsc_b  hsc_s  degree_p  degree_t workex  \
0       0  67.00      1  91.00      0      0     58.00         1     No   
1       0  79.33      0  78.33      0      1     77.48         1    Yes   
2       0  65.00      0  68.00      1      2     64.00         0     No   
3       0  56.00      0  52.00      1      1     52.00         1     No   
4       0  85.80      0  73.60      1      0     73.30         0     No   

   etest_p  specialisation  mba_p  status  
0     55.0               1  58.80       1  
1     86.5               0  66.28       1  
2     75.0               0  57.80       1  
3     66.0               1  59.43       0  
4     96.8               0  55.50       1

In [36]:
data_edited['workex'].value_counts()

No     141
Yes     74
Name: workex, dtype: int64

In [37]:
data_edited['workex'] = data_edited['workex'].map({'Yes' : 1,'No' : 0})

In [40]:
data_edited.isna().sum()

gender            0
ssc_p             0
ssc_b             0
hsc_p             0
hsc_b             0
hsc_s             0
degree_p          0
degree_t          0
workex            0
etest_p           0
specialisation    0
mba_p             0
status            0
dtype: int64

In [41]:
data_edited.describe()

gender       ssc_p       ssc_b       hsc_p       hsc_b       hsc_s  \
count  215.000000  215.000000  215.000000  215.000000  215.000000  215.000000   
mean     0.353488   67.303395    0.460465   66.333163    0.390698    0.525581   
std      0.479168   10.827205    0.499598   10.897509    0.489045    0.594403   
min      0.000000   40.890000    0.000000   37.000000    0.000000    0.000000   
25%      0.000000   60.600000    0.000000   60.900000    0.000000    0.000000   
50%      0.000000   67.000000    0.000000   65.000000    0.000000    0.000000   
75%      1.000000   75.700000    1.000000   73.000000    1.000000    1.000000   
max      1.000000   89.400000    1.000000   97.700000    1.000000    2.000000   

         degree_p    degree_t      workex     etest_p  specialisation  \
count  215.000000  215.000000  215.000000  215.000000      215.000000   
mean    66.370186    0.376744    0.344186   72.100558        0.441860   
std      7.358743    0.581987    0.476211   13.275956        0.497767   
min     50.000000    0.000000    0.000000   50.000000        0.000000   
25%     61.000000    0.000000    0.000000   60.000000        0.000000   
50%     66.000000    0.000000    0.000000   71.000000        0.000000   
75%     72.000000    1.000000    1.000000   83.500000        1.000000   
max     91.000000    2.000000    1.000000   98.000000        1.000000   

            mba_p      status  
count  215.000000  215.000000  
mean    62.278186    0.688372  
std      5.833385    0.464240  
min     51.210000    0.000000  
25%     57.945000    0.000000  
50%     62.000000    1.000000  
75%     66.255000    1.000000  
max     77.890000    1.000000

In [42]:
data_used = data_edited.copy()

In [63]:
from sklearn.model_selection import train_test_split
train_data, test_data = train_test_split(data_used, test_size=0.2)

In [64]:
train_data.shape

(172, 13)

In [65]:
test_data.shape

(43, 13)

In [66]:
from sklearn import preprocessing
train_normalize = preprocessing.normalize(train_data.drop('status',axis=1))
test_normalize = preprocessing.normalize(test_data.drop('status',axis=1))

In [67]:
train_normalize

array([[0.        , 0.42065785, 0.        , ..., 0.53689226, 0.00553497,
        0.39010481],
       [0.        , 0.41024333, 0.        , ..., 0.55581354, 0.        ,
        0.42446951],
       [0.        , 0.43489487, 0.        , ..., 0.40144142, 0.00669069,
        0.41361848],
       ...,
       [0.00661523, 0.50275766, 0.        , ..., 0.38368348, 0.00661523,
        0.44011141],
       [0.        , 0.46496955, 0.00749951, ..., 0.51326638, 0.00749951,
        0.47074417],
       [0.00678118, 0.46790122, 0.00678118, ..., 0.45433887, 0.00678118,
        0.4228064 ]])

In [68]:
test_normalize

array([[0.        , 0.48077739, 0.00552618, 0.48077739, 0.        ,
        0.        , 0.37578003, 0.        , 0.        , 0.52498681,
        0.00552618, 0.34759653],
       [0.00549942, 0.46107158, 0.00549942, 0.49401312, 0.        ,
        0.        , 0.42455542, 0.        , 0.00549942, 0.43302453,
        0.        , 0.41894601],
       [0.        , 0.42592162, 0.0059989 , 0.47391279, 0.        ,
        0.        , 0.39592714, 0.        , 0.0059989 , 0.56389623,
        0.        , 0.34523647],
       [0.00630192, 0.49028941, 0.        , 0.40332291, 0.00630192,
        0.00630192, 0.40458329, 0.00630192, 0.        , 0.47579499,
        0.00630192, 0.45462054],
       [0.        , 0.42142692, 0.00690864, 0.48360466, 0.        ,
        0.        , 0.44215283, 0.        , 0.        , 0.47324171,
        0.00690864, 0.41106396],
       [0.        , 0.334873  , 0.00683414, 0.40321443, 0.        ,
        0.00683414, 0.44421929, 0.00683414, 0.00683414, 0.58773629,
        0.        ,

In [69]:
x_train = train_normalize.copy()
y_train = train_data['status']
x_test = test_normalize.copy()
y_test = test_data['status']

In [70]:
x_train.shape

(172, 12)

In [71]:
y_train.shape

(172,)

In [72]:
x_test.shape

(43, 12)

In [73]:
y_test.shape

(43,)

### How to do Regression?

- Logistic Regression
- Support Vector Machine
- Random Forest
- Neural Networks
- Decision Tree
- K-Nearest Neighbors
- Gaussian Naive Bayes
- Linear Discriminant Analysis

### Logistic Regression

In [53]:
from sklearn.linear_model import LogisticRegression
LR = LogisticRegression()

In [75]:
LR.fit(x_train,y_train)
print('Accuracy of Logistic Regression on Training set : {:.2f}'.format(LR.score(x_train,y_train)))

Accuracy of Logistic Regression on Training set : 0.70


### Support Vector Machine

In [76]:
from sklearn import svm
svm = svm.LinearSVC()

In [77]:
svm.fit(x_train,y_train)
print('Accuracy of Support Vector Machine on Training Set: {:.2f}'.format(svm.score(x_train,y_train)))

Accuracy of Support Vector Machine on Training Set: 0.78


### Random Forest

In [78]:
from sklearn.ensemble import RandomForestClassifier
RF = RandomForestClassifier()

In [80]:
RF.fit(x_train,y_train)
print('Accuracy of Random Forest on Training Set : {:.2f}'.format(RF.score(x_train,y_train)))

Accuracy of Random Forest on Training Set : 1.00


### Neural Networks

In [105]:
from sklearn.neural_network import MLPClassifier
NN = MLPClassifier(max_iter=650)

In [106]:
NN.fit(x_train,y_train)
print('Accuracy of Neural Networks on Training Set : {:.2f}'.format(NN.score(x_train,y_train)))

Accuracy of Neural Networks on Training Set : 0.70


Perlu di gridsearch

### Decision Tree

In [84]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier()

In [85]:
clf.fit(x_train,y_train)
print('Accuracy of Decision Tree Classifier on Training Set : {:.2f}'.format(clf.score(x_train,y_train)))

Accuracy of Decision Tree Classifier on Training Set : 1.00


### K-nearest Neighbor

In [107]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()

In [108]:
knn.fit(x_train,y_train)
print('Accuracy of K-Nearest Neighbors on Training Set : {:.2f}'.format(knn.score(x_train,y_train)))

Accuracy of K-Nearest Neighbors on Training Set : 0.88


### Gaussian Naive Bayes

In [109]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

In [110]:
gnb.fit(x_train,y_train)
print('Accuracy of Gaussian Naive Bayes on Training set : {:.2f}'.format(gnb.score(x_train,y_train)))

Accuracy of Gaussian Naive Bayes on Training set : 0.90


### Linear Discriminant Analysis

In [111]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
lda = LinearDiscriminantAnalysis()

In [112]:
lda.fit(x_train,y_train)
print('Accuracy of Linear Discriminant Analysis on Training set : {:.2f}'.format(lda.score(x_train,y_train)))

Accuracy of Linear Discriminant Analysis on Training set : 0.90


### Which are the best Classifier, based on accuracy on Training set?
1. Random Forest & Decision Tree (1.00)
2. Gaussian Naive Bayes & Linear Discriminant Analysis (0.90)
3. K-Nearest Neighbours (0.88)
4. SVM (0.78)
5. Logistic Regression & Neural Networks (0.70)

I''ll use random forest, decision tree, Gaussian Naive Bayes, and Linear Discriminant analysis for the next step. Because 4 of those has the best accuracy, but not yet tested on test dataset.